Fix Clean Up Palette Related Crash and cleanup parameters related adjustment

Abel032 opened the pull request:
### How to reproduce the crash

1. Open a scene
2. Save All
3. switch cleanup related opinions(like change mono mode to color mode)
4. a palette related crash

This PR would resolve the crash that related to a dangling palette pointer and a nullptr of palette.

To not store current cleanup parameters to scene's parameters' cache(already write to scene file) if autosave.

Fixes #5257

Abel032 committed:
Fix Clean Up Crash

Abel032 commented:
This PR caused a Save delay.
Consider another way to avoid dangling pointer.

RodneyBaker commented:
We need  a plan to refine the Save process although that is outside the scope of this PR.  In the long run use of std::shared_ptr or even std::weak_ptr could be used (in the later case to prevent potential circular reference/memory leak.  In other words to reference the palette controller but not own it.

Translation:  I'm not sure how to best proceed here but I am glad you are taking an active look into it.

Abel032 commented:
I'm not quite understand the relationship between GlobalParameters(for project?) and CurrentParameters(for scene).

![图片](https://github.com/user-attachments/assets/ee70cad5-1c26-4789-81a9-cfc9c4f1fe17)

The GlobalParameters always assign from CurrentParameters, but CurrentParameters only assign from GlobalParameters in two place:
CleanupSettingsModel::restoreGlobalSettings()
and
IoCmd::saveScene(const TFilePath &path, int flags)

Abel032 committed:
Remove some References of GlobalParameters

Abel032 committed:
Fix Cleanup Palette Name in StyleEditor

Abel032 commented:
Let's see if it works..

Abel032 commented:
There might be something wrong, but I can't reproduce the crash now.

andeon commented:
I’m not sure, but this pull request (https://github.com/tahoma2d/tahoma2d/pull/1605) probably resolves the palette switching issue. If I remember correctly, this issue also occurred in older versions of Tahoma.

Abel032 commented:
@andeon 
Yes I have checked that PR before, they do have some like, but I couldn't understand the reason that caused crash in that PR.
Because I can also reproduce the crash with tahoma now.(but steps are different through)

But I guess these crash all related to dangling palette pointers.

I'll check that PR one more time.Thanks.

Abel032 committed:
Fix nullptr of Level's Palette when getting icon of Column

Abel032 committed:
Update Parameters when Save Scene by user actively

Abel032 commented:
This PR has through my test now.
Nothing to commit if no more crash.

RodneyBaker commented:
It looks like in this process you resolved an Autosave issue.
Nice.

That's an area of interest that would be nice to look at with regard to the actual autosave code so that any area touched by autosave doesn't run into a similar problem.  This as opposed to implementing autosave fixes throughout the source code.

Abel032 commented:
> It looks like in this process you resolved an Autosave issue.

Little adjust, to not store current cleanup parameters to scene's parameters' cache(already write to scene file) if autosave.

Abel032 committed:
Add a comment

// Don't store current cleanup parameters to scene's parameters' cache if autosave
// (would save to scene file) .

Abel032 committed:
Rename and update CleanupSettingsPane restore function

Abel032 committed:
Merge branch 'opentoonz:master' into Fix-CleanUp-Crash

Abel032 commented:
> That's an area of interest that would be nice to look at with regard to the actual `autosave` code so that any area touched by `autosave` doesn't run into a similar problem. This as opposed to implementing `autosave` fixes throughout the source code.

I think the `autosave` process won't be problem, it's basically the regular repeat of `SaveAll` (or save only resouces).

Before this PR, the save scene process would try to copy the `GlobalCleanupParams` and replace the `CurrentCleanupParams`,
and as a palette pointer in the `CurrentCleanupParams` which is copied to `PaletteController` for quick switch, it becomes a dangling pointer.

The save scene process and some cleanup related behavior would change current palette to cleanup palette, this caused the crash.

Fix the management of cleanup palette pointer(which means the management of two `CleanupParams`) would fix these crashes.

Abel032 commented:
And propbly fix #5296 
For those who want to test, here are the portable versions:
Windows:
[Download Here](https://github.com/opentoonz/opentoonz/actions/runs/13227017892)
MacOS :
[Download Here](https://github.com/opentoonz/opentoonz/actions/runs/13227017893)

RodneyBaker commented:
Posting intent to merge.
From all appearances this fix works well and it will be good get the fixe to more people for field testing.

I don't think we've solved all issues related to cleanup and crashing yet but this will move that effort forward.  

After merging of this PR, a follow up by the community identifying cleanup processes still problematic will be appreciated.

RodneyBaker commented:
Thank you @Abel032 for the fix.
